File: src/main/java/com/example/store/security/DomainUserDetailsService.java
package com.example.store.security;

import com.example.store.data.User;
import com.example.store.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;

@Component("userDetailsService")
public class DomainUserDetailsService implements UserDetailsService {

    @Autowired
    private UserRepository userRepository;

    @Override
    public UserDetails loadUserByUsername(String s) throws UsernameNotFoundException {
        return userRepository.findByEmail(s)
                .map(user -> createSpringSecurityUser(s, user))
                .orElseThrow(() -> new BadCredentialsException("Invalid Username Or Password"));
    }

    private UserDetails createSpringSecurityUser(String lowercaseLogin, User user) {
        List<GrantedAuthority> grantedAuthorities = new ArrayList<>();
        grantedAuthorities.add(new SimpleGrantedAuthority(user.getAuthorities().getName()));
        return new org.springframework.security.core.userdetails.User(lowercaseLogin, user.getPassword(), grantedAuthorities);
    }
}

File: src/main/kotlin/com/example/store/security/DomainUserDetailsService.kt
package com.example.store.security

import com.example.store.data.User
import com.example.store.repository.UserRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Component

@Component("userDetailsService")
class DomainUserDetailsService : UserDetailsService {

    @Autowired
    private lateinit var userRepository: UserRepository

    @Throws(UsernameNotFoundException::class)
    override fun loadUserByUsername(s: String?): UserDetails? {
        return userRepository.findByEmail(s)
            ?.map { user: User? -> createSpringSecurityUser(s, user) }
            ?.orElseThrow { BadCredentialsException("Invalid Username Or Password") }
    }

    private fun createSpringSecurityUser(lowercaseLogin: String?, user: User?): UserDetails {
        val grantedAuthorities: MutableList<GrantedAuthority?> = ArrayList()
        if (user != null) {
            grantedAuthorities.add(SimpleGrantedAuthority(user.getAuthorities()?.getName() ?: ""))
            return org.springframework.security.core.userdetails.User(
                lowercaseLogin,
                user.getPassword(),
                grantedAuthorities
            )
        } else {
            return org.springframework.security.core.userdetails.User("", "", grantedAuthorities)
        }
    }
}